sn_client/src/folders.rs

```diff
@@ -99,32 +140,67 @@ impl FoldersApi {
         address: RegisterAddress,
     ) -> Result<Self> {
         let register = ClientRegister::retrieve(client.clone(), address).await?;
-        Ok(Self {
-            client,
-            wallet_dir: wallet_dir.to_path_buf(),
-            register,
-        })
+        Self::create(client, wallet_dir, register)
     }
 
     /// Returns the list of entries of this Folder
-    pub fn entries(&self) -> Result<Vec<(OsString, FolderEntry)>> {
+    pub async fn entries(&mut self) -> Result<Vec<Metadata>> {
         let mut entries = vec![];
         for (_, entry) in self.register.read() {
-            let (name, folder_entry): (String, FolderEntry) = rmp_serde::from_slice(&entry)?;
-            entries.push((name.into(), folder_entry));
+            let mut xorname = [0; XOR_NAME_LEN];
+            xorname.copy_from_slice(&entry);
+            let metadata_addr = XorName(xorname);
+            let metadata = match self.metadata.get(&metadata_addr) {
+                Some((metadata, _)) => metadata.clone(),
+                None => {
+                    // retrieve metadata Chunk from network
+                    let chunk = self
+                        .client
+                        .get_chunk(ChunkAddress::new(metadata_addr), false, None)
+                        .await?;
+                    let metadata: Metadata = rmp_serde::from_slice(chunk.value())?;
+                    self.metadata
+                        .insert(metadata_addr, (metadata.clone(), None));
+                    metadata
+                }
+            };
+            entries.push(metadata);
         }
         Ok(entries)
     }
 
     // Private helpers
 
-    // Add the given entry to the underlying Register
-    fn add_entry(&mut self, name: OsString, entry: FolderEntry) -> Result<()> {
-        // TODO: conversion to String will be removed when metadata is moved out of the Register
-        let name = name.into_string().unwrap_or("unknown".to_string());
-        let entry = (name, entry);
+    // Create a new FoldersApi instance with given register.
+    fn create(client: Client, wallet_dir: &Path, register: ClientRegister) -> Result<Self> {
+        let files_api = FilesApi::new(client.clone(), wallet_dir.to_path_buf());
+
+        Ok(Self {
+            client,
+            wallet_dir: wallet_dir.to_path_buf(),
+            register,
+            files_api,
+            metadata: BTreeMap::new(),
+        })
+    }
+
+    // Add the given entry to the underlying Register as well as creating the metadata Chunk
+    fn add_entry(&mut self, name: OsString, content: FolderEntry) -> Result<(Metadata, XorName)> {
+        // create metadata Chunk for this entry
+        let metadata = Metadata {
+            name: name.to_str().unwrap_or("unknown").to_string(),
```

Comment:
Should this be just the `OsString`? The previous todo said the following
`// TODO: conversion to String will be removed when metadata is moved out of the Register`

Reply:
there is a problem with wasm build and unsafe api fom osstring, it needs more investigation